modules/core/src/test/java/org/apache/ignite/internal/processors/cache/datastructures/IgniteReentrantLockTest.java

```diff
@@ -0,0 +1,314 @@
+import org.apache.ignite.configuration.AtomicConfiguration;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.internal.IgniteInternalFuture;
+import org.apache.ignite.spi.discovery.tcp.TcpDiscoverySpi;
+import org.apache.ignite.spi.discovery.tcp.ipfinder.TcpDiscoveryIpFinder;
+import org.apache.ignite.spi.discovery.tcp.ipfinder.vm.TcpDiscoveryVmIpFinder;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+
+/** */
+public class IgniteReentrantLockTest extends GridCommonAbstractTest {
+    /** */
+    private static final TcpDiscoveryIpFinder ipFinder = new TcpDiscoveryVmIpFinder(true);
+
+    /** */
+    private static final int NODES_CNT = 4;
+
+    /** */
+    private static final Random RND = new Random();
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        TcpDiscoverySpi spi = new TcpDiscoverySpi();
+
+        spi.setIpFinder(ipFinder);
+
+        cfg.setDiscoverySpi(spi);
+
+        AtomicConfiguration atomicCfg = new AtomicConfiguration();
+
+        assertNotNull(atomicCfg);
+
+        atomicCfg.setCacheMode(CacheMode.REPLICATED);
+
+        cfg.setAtomicConfiguration(atomicCfg);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        startGridsMultiThreaded(NODES_CNT);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testInitialization() throws Exception {
+        try (IgniteLock lock = createReentrantLock(0, "lock1", false)) {
+        }
+
+        try (IgniteLock lock = createReentrantLock(0, "lock2", true)) {
+        }
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLock() throws Exception {
+        testReentrantLock(true);
+        testReentrantLock(false);
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLock(final boolean fair) throws Exception {
+        for (int i = 0; i < NODES_CNT; i++) {
+            IgniteLock lock = createReentrantLock(i, fair ? "lock1" : "lock2", fair);
+
+            // Other node can still don't see update.
+            while (lock.isLocked())
+                ;
+
+            lock.lock();
+
+            assertTrue(lock.isLocked());
+
+            lock.lock();
+
+            assertTrue(lock.isLocked());
+
+            lock.unlock();
+
+            assertTrue(lock.isLocked());
+
+            lock.unlock();
+
+            assertFalse(lock.isLocked());
+        }
+    }
+
+    /**
+     * @throws Exception If failed.
+     */
+    public void testReentrantLockMultinode() throws Exception {
```

Comment:
javadoc with descriptions for test methods)

Reply:
fixed